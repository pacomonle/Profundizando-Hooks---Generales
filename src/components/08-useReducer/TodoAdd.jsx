import React from 'react'
import useForm from '../../hooks/useForm'
import './styles.css'

const TodoAdd = ({ handleAddTodo}) => {

    // hook personalizado useForm para el input del form
const [formValue, reset,  handleInputChange] = useForm({
    inputValue: ''
})

const  {inputValue} = formValue

const handleSubmit = (e) => {
    e.preventDefault();
if( inputValue.trim().length < 1){ return }
    const newTodo = {
        id: new Date().getTime(),
        desc: inputValue,
        done: false
    }

handleAddTodo(newTodo)

reset()
}

    return (
        <>
        <h3>Agregar TODO</h3>
        <hr />
        <form onSubmit={handleSubmit}>
            <input 
            type="text" 
            className='form-control'
            placeholder='Aprender...'
            autoComplete= 'off'
            name="inputValue" 
            value= {inputValue}
            onChange={handleInputChange} 
            />
            <button 
            className='btn btn-outline-primary btn-block my-1'
            type="submit">
            Agregar
            </button>
        </form>
        </>
    )
}

export default TodoAdd
