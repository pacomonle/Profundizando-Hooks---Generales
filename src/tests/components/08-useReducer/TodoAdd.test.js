import React from 'react';
import { shallow } from 'enzyme';
import TodoAdd  from '../../../components/08-useReducer/TodoAdd';

describe('Pruebas en <TodoAdd />', () => {
 // para simular las funnction de las props
    const handleAddTodo = jest.fn();

    const wrapper = shallow( 
        <TodoAdd 
            handleAddTodo={ handleAddTodo }
        /> 
    )


    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('NO debe de llamar handleAddTodo', () => {
        
        // comprobar la validacion cuando no enviamos nada al submit  
        const formSubmit = wrapper.find('form').prop('onSubmit');
        // console.log(formSubmit)
        // simulacion del evento sin usar el simulate
        formSubmit({ preventDefault(){} });

        expect( handleAddTodo ).toHaveBeenCalledTimes(0);


    });

    test('debe de llamar la funciÃ³n handleAddTodo', () => {
    // simulando en el input pra enviar un argumento al submit
        const value = 'Aprender Firestore';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'inputValue'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });

        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) ); // { } event del input
        expect( handleAddTodo ).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        }); 

       // comprobar que se limpia el input al salir del submit
        expect( wrapper.find('input').prop('value') ).toBe('');




    })
    
    

    

    
})
