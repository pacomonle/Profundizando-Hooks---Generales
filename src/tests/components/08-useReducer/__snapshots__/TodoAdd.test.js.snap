// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    Agregar TODO
  </h3>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="inputValue"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block my-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
